Mathlib/RingTheory/Localization/Algebra.lean

```diff
@@ -0,0 +1,173 @@
+lemma RingHom.toKerIsLocalization_isLocalizedModule (hT : Submonoid.map g M = T) :
+    IsLocalizedModule M (toKerIsLocalization S Q g (hT.symm ▸ Submonoid.le_comap_map M)) where
+  map_units x := by
+    rw [isUnit_iff_exists]
+    have hu : IsUnit (algebraMap R S x) := IsLocalization.map_units S x
+    let φ : Module.End R
+        (RingHom.ker (IsLocalization.map Q g (hT.symm ▸ Submonoid.le_comap_map M) : S →+* Q)) := {
+      toFun := fun y ↦ ⟨hu.unit⁻¹ * y, by
+        simp [RingHom.mem_ker, (RingHom.mem_ker _).mp y.property]⟩
+      map_add' := fun x y ↦ by simp [mul_add]
+      map_smul' := fun a x ↦ by simp
+    }
+    refine ⟨φ, ?_, ?_⟩
+    · ext y
+      simp [φ, Algebra.smul_def, ← mul_assoc]
+    · ext y
+      simp [φ, Algebra.smul_def, ← mul_assoc]
+  surj' y := by
+    subst hT
+    obtain ⟨⟨r, m⟩, hx⟩ := IsLocalization.surj (M := M) y.val
+    rw [mul_comm] at hx
+    have heq : algebraMap P Q (g r) = algebraMap P Q 0 := by
+      rw [← IsLocalization.map_eq (S := S) (Submonoid.le_comap_map M), ← hx]
+      simp [(RingHom.mem_ker _).mp y.property]
+    obtain ⟨⟨_, t, tM, rfl⟩, ht⟩ := (IsLocalization.eq_iff_exists (Submonoid.map g M) _).mp heq
+    simp only [mul_zero] at ht
+    have hr : t * r ∈ RingHom.ker g := by
+      simp only [RingHom.mem_ker, map_mul]
+      exact ht
+    refine ⟨⟨⟨t * r, hr⟩, ⟨t, tM⟩ * m⟩, Subtype.ext ?_⟩
+    rw [SetLike.val_smul_of_tower, toKerIsLocalization_apply, Submonoid.smul_def, Algebra.smul_def]
+    simp [mul_assoc, hx]
+  exists_of_eq {x y} h := by
+    obtain ⟨c, hc⟩ := IsLocalization.exists_of_eq (M := M) (congrArg Subtype.val h)
+    exact ⟨c, Subtype.ext hc⟩
+
+section Algebra
+
+variable {R : Type u} [CommRing R] (M : Submonoid R)
+variable {A : Type v} [CommRing A] [Algebra R A]
+variable {B : Type w} [CommRing B] [Algebra R B]
+variable {Rₚ : Type u'} [CommRing Rₚ] [Algebra R Rₚ] [IsLocalization M Rₚ]
+variable {Aₚ : Type v'} [CommRing Aₚ] [Algebra R Aₚ] [Algebra A Aₚ] [IsScalarTower R A Aₚ]
+  [IsLocalization (Algebra.algebraMapSubmonoid A M) Aₚ]
+variable {Bₚ : Type v'} [CommRing Bₚ] [Algebra R Bₚ] [Algebra B Bₚ] [IsScalarTower R B Bₚ]
+  [IsLocalization (Algebra.algebraMapSubmonoid B M) Bₚ]
+variable [Algebra Rₚ Aₚ] [Algebra Rₚ Bₚ] [IsScalarTower R Rₚ Aₚ] [IsScalarTower R Rₚ Bₚ]
+
+namespace IsLocalization
+
+lemma map_algebraMapSubmonoid_eq (f : A →ₐ[R] B) :
+    (Algebra.algebraMapSubmonoid A M).map f = Algebra.algebraMapSubmonoid B M := by
+  ext x
+  constructor
+  · rintro ⟨a, ⟨r, hr, rfl⟩, rfl⟩
+    simp only [AlgHom.commutes]
+    use r
+  · rintro ⟨r, hr, rfl⟩
+    simp only [Submonoid.mem_map]
+    use (algebraMap R A r)
+    simp only [AlgHom.commutes, and_true]
+    use r
+
+lemma algebraMapSubmonoid_le_comap (f : A →ₐ[R] B) :
+    Algebra.algebraMapSubmonoid A M ≤ (Algebra.algebraMapSubmonoid B M).comap f.toRingHom := by
+  rw [← map_algebraMapSubmonoid_eq M f]
+  exact Submonoid.le_comap_map (Algebra.algebraMapSubmonoid A M)
```

Comment:
Are there better places for these to go?

Reply:
I moved them.